### PR TITLE
Drop illuminate/support requirement

### PR DESCRIPTION
**Pull request by Jimlah**
#86

### CODE REVIEW

**Comment by Jimlah** [comment]
@Gummibeer Here is the PR for the #86

**Comment by Sammyjo20** [comment]
Thanks @Jimlah I will review this weekend. Don't worry if the code style fails.

File: composer.json
Comment (Gummibeer) [inline composer.json]: as `9.34` isn't really required but `^9.0` works as well I would go for that constraint.
PS: if it's only for the `collect()` method it would be [`illuminate/collections`](https://github.com/illuminate/collections) instead of the whole `illuminate/support` package.

Comment (Jimlah) [reply composer.json]: I will work on it and send a new PR

Comment (Gummibeer) [reply composer.json]: Don't send a new PR - just commit to your already existing branch. It will update this PR. 😉

Comment (Jimlah) [reply composer.json]: Done

File: tests/Feature/Oauth2/AuthCodeFlowConnectorTest.php
Comment (Sammyjo20) [inline tests/Feature/Oauth2/AuthCodeFlowConnectorTest.php:53]: I can't see where this method has been defined, is this a typo? Or is this a built-in pest function?

File: src/Http/SaloonResponse.php
Comment (Sammyjo20) [inline src/Http/SaloonResponse.php:196]: Would we also be able to add a custom message to this to suggest installing `illuminate/collection` please? We may need our own exception for this. What do you think?

File: src/Http/SaloonResponse.php
Comment (Sammyjo20) [inline src/Http/SaloonResponse.php:157]: Ah, I actually think we're losing a bit of functionality if we switch back to using normal array access. With `Arr::get` you can do cool things like `nestedObject.value` and it will return the value from the dot-notion. Is it possible if we could extract Arr::get into our own helper to keep this logic?

File: src/Traits/CollectsConfig.php
Comment (Sammyjo20) [inline src/Traits/CollectsConfig.php:89]: Same comment regarding Arr::get on src/Http/SaloonResponse.php

File: src/Traits/CollectsData.php
Comment (Sammyjo20) [inline src/Traits/CollectsData.php:104]: Same comment regarding Arr::get on src/Http/SaloonResponse.php

File: src/Traits/CollectsHeaders.php
Comment (Sammyjo20) [inline src/Traits/CollectsHeaders.php:88]: Same comment regarding Arr::get on src/Http/SaloonResponse.php

File: src/Traits/CollectsQueryParams.php
Comment (Sammyjo20) [inline src/Traits/CollectsQueryParams.php:100]: Same comment regarding Arr::get on src/Http/SaloonResponse.php

File: src/Helpers/URLHelper.php
Comment (Sammyjo20) [inline src/Helpers/URLHelper.php:60]: How come we're not using the same code as Illuminate/support Str here? https://github.com/illuminate/support/blob/master/Str.php#L357

File: src/Helpers/StateHelper.php
Comment (Sammyjo20) [inline src/Helpers/StateHelper.php:26]: Same here, I kind of wish we had the same functionality as: https://github.com/illuminate/support/blob/master/Str.php#L726

**Review by Sammyjo20 — CHANGES_REQUESTED** [review]
Hey, thanks for your PR - it's much appreciated.

Apologies for lots of comments, and great work but I'm having a few thoughts and I'm not sure if I necessarily want to remove Illuminate/Support from Saloon. There are so many helper methods that are useful and while it's great to have alternatives inside Saloon I feel it's extra code to continuously maintain when Laravel and the core team will likely continue maintaining illuminate/support for a very long time. 

I kind of wish there were smaller helpers where we could bring in just "String helpers" or just "collections" but I don't think this is a thing.

So in summary, great PR but I think it's worth discussing if it's worth introducing this much code into the core just to remove a dependant?

**Comment by Sammyjo20** [comment]
@Gummibeer Tagging you for your thoughts too.

**Comment by Sammyjo20** [comment]
I've re-read the issue that you posted on #86 and I understand a bit more now why it's useful to remove it.

**Comment by Jimlah** [comment]
While working on it. I did wished we could have left the illuminate support, because of the many helper function it supported. But then again I could see that along that the helper function where not fully used. And about the collection, most non laravel users may not fully know about it. Untill they read about it.

File: tests/Feature/Oauth2/AuthCodeFlowConnectorTest.php
Comment (Gummibeer) [reply tests/Feature/Oauth2/AuthCodeFlowConnectorTest.php:53]: It's a native PHP function. 
https://www.php.net/manual/de/function.str-ends-with.php

**Comment by Gummibeer** [comment]
Regarding the reason to remove it: the support package is one of the fundamental base packages of Laravel framework. Even the support package itself has a bunch of dependencies - following as of writing:
```
        https://github.com/doctrine/inflector: "^2.0",
        https://github.com/illuminate/collections: "^10.0",
        https://github.com/illuminate/conditionable: "^10.0",
        https://github.com/illuminate/contracts: "^10.0",
        https://github.com/illuminate/macroable: "^10.0",
        https://github.com/briannesbitt/Carbon: "^2.53.1",
        https://github.com/voku/portable-ascii: "^2.0"
```

These alone will quickly sum up to a pretty big tree of dependencies that can conflict with the package itself and/or the plain PHP users of the package. Laravel also has a pretty modern PHP requirement all the time. So I know it's the ideal world but the reality is that a lot of PHP projects, primarily the plain PHP ones, stay on PHP7.4 for decades.

But also for Laravel users: you will have to upgrade it at least once per year and decide if you only add the new version or drop the old one. Adding will make it easier for users and is likely what's needed as the used helpers don't really change since years. But this will increase your maintenance work. As you have to support the whole tree of dependencies for every Laravel release. And this will possibly break your neck. As soon as a new PHPUnit major version or PHP major version is released and used by Laravel you will have a painful time.

I've been there with @astrotomic / laravel-translatable to support as many PHP/Laravel versions as possible. It's absolutely painful! And for a plain PHP package having this trouble is absolutely not needed.

**Comment by Sammyjo20** [comment]
Thank you for your effort @Jimlah however I've decided that I'm going to hold off on these changes until v2, since it requires breaking changes to remove the Carbon package and Illuminate Support package, e.g for the `collect` method.

**Pull request closed without merging**
